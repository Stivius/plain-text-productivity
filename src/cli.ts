#! /usr/bin/env node

import main from "./index";

main().catch((err) => console.log(err, "Error while running"));
