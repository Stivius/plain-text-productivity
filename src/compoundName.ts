const SEPARATOR = "/";

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundName {
  parent: CompoundName;
  name: string;

  constructor(value: string) {
    const splitted = value.split(SEPARATOR);
    if (splitted.length > 0) {
      const name = splitted.pop();
      if (name.length === 0) {
        throw Error("subpart cannot be empty");
      }
      if (splitted.length > 0) {
        this.parent = new CompoundName(splitted.join(SEPARATOR));
      }
      this.name = name;
    } else {
      this.name = value;
    }
  }

  toString(): string {
    if (this.parent != null) {
      return `${this.parent.toString()}/${this.name}`;
    }
    return this.name;
  }

  isEqual(other: CompoundName): boolean {
    return this.toString() === other.toString();
  }
}

export const searchForCompoundProject = (searchedValue: CompoundName) => {
  return (value: CompoundName) => value.isEqual(searchedValue);
};
