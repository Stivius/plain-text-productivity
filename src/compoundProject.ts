const SEPARATOR = '/'

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundProject {
    parent: CompoundProject;
    name: string;

    constructor(value: string) {
        const splitted = value.split(SEPARATOR);
        if (splitted.length > 0) {
            const name = splitted.pop();
            if (name.length === 0) {
                throw Error('subpart cannot be empty');
            }
            if (splitted.length > 0) {
                this.parent = new CompoundProject(splitted.join(SEPARATOR));
            }
            this.name = name;
        } else {
            this.name = value;
        }
    }

    toString(): string {
        if (this.parent != null) {
            return `${this.parent.toString()}/${this.name}`
        }
        return this.name;
    }

    isEqual(other: CompoundProject): boolean {
        return this.toString() === other.toString();
    }
}
