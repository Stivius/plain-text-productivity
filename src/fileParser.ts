import fs from 'fs';
import moment from 'moment';
import {
    FileData,
    ProjectRecord,
    Mark,
    EstimatedProject,
    DATE_REGEX,
    PROJECT_NAME_REGEX,
    METADATA_PROJECT_NAME_REGEX
} from "./interfaces";

enum ParsingState {
    Init = 'init',
    MetadataStarted = 'metadata_started',
    MetadataFinished = 'metadata_finished',
    MetadataActiveProjectsStarted = 'metadata_active_projects_started',
    MetadataActiveProjects = 'metadata_active_projects',
    MetadataCompletedProjects = 'metadata_completed_projects',
    ArchivedProjects = 'archived_projects',
    Record = 'record'
}

type StateAction = (line: string) => void;
type StateTransitionCondition = (line: string) => boolean;

interface Transition {
    newState: ParsingState;
    condition: StateTransitionCondition;
}

class ParsingStateMachine {
    private state = ParsingState.Init;
    private stateActions = new Map<ParsingState, StateAction>();
    private stateTransitions = new Map<ParsingState, Transition[]>();

    addStateAction(state: ParsingState, action: StateAction) {
        this.stateActions.set(state, action);
    }

    addStateTransition(state: ParsingState, newState: ParsingState, condition: StateTransitionCondition) {
        if (!this.stateTransitions.has(state)) {
            this.stateTransitions.set(state, []);
        }
        let transitions = this.stateTransitions.get(state);
        transitions.push({ 
            newState: newState,
            condition: condition
        });
    }

    parseLine(line: string) {
        const transitions = this.stateTransitions.get(this.state);
        if (transitions !== undefined) {
            for (const transition of transitions) {
                if (transition.condition(line)) {
                    this.state = transition.newState;  
                    break;
                }
            }
        }
        this.performActionForState(line);
    }
    
    private performActionForState(line: string) {
        const action = this.stateActions.get(this.state);
        if (action !== undefined) {
            action(line);
        }
    }
}

export class FileParser {
    private projects: string[] = [];
    private records: ProjectRecord[] = [];
    private currentRecord: ProjectRecord = undefined;

    constructor(private filename: string) { }

    parse(): FileData {
        const data = fs.readFileSync(this.filename);
        const lines = data.toString().split('\n');
        const stateMachine = new ParsingStateMachine();

        stateMachine.addStateAction(
            ParsingState.Record,
            (line: string) => {
                if (this.isStartOfNewRecord(line)) {
                    if (this.currentRecord !== undefined) {
                        this.records.push(this.currentRecord);
                    }
                    this.currentRecord = this.createRecord(line);
                } else {
                    this.currentRecord.projects.push(this.parseProject(line));
                }
            }
        );

        stateMachine.addStateAction(
            ParsingState.MetadataActiveProjects,
            (line: string) => {
                this.projects.push(this.parseMetadataProject(line));
            }
        );

        stateMachine.addStateTransition(
            ParsingState.Init,
            ParsingState.MetadataStarted,
            (line: string) => {
                return this.isMetadataSeparator(line);
            }
        );

        stateMachine.addStateTransition(
            ParsingState.Init,
            ParsingState.Record,
            (line: string) => {
                return this.isStartOfNewRecord(line);
            }
        );

        stateMachine.addStateTransition(
            ParsingState.MetadataStarted,
            ParsingState.MetadataActiveProjectsStarted,
            (line: string) => {
                return this.isStartOfActiveProjects(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataActiveProjectsStarted,
            ParsingState.MetadataActiveProjects,
            (line: string) => {
                return true;
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataActiveProjects,
            ParsingState.MetadataFinished,
            (line: string) => {
                return this.isMetadataSeparator(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataFinished,
            ParsingState.Record,
            (line: string) => {
                return this.isStartOfNewRecord(line);
            }
        )

        for (const line of lines) {
            const trimmedLine = line.trim();
            if (trimmedLine.length === 0)
                continue;

            stateMachine.parseLine(trimmedLine);
        }
        if (this.currentRecord !== undefined) {
            this.records.push(this.currentRecord);
        }
        return { metadata: { projects: this.projects }, records: this.records };
    }

    private isMetadataSeparator(line: string) {
        return line === '---';
    }

    private isStartOfNewRecord(line: string) {
        return line.match(DATE_REGEX) !== null;
    }

    private isStartOfActiveProjects(line: string) {
        return line.indexOf('Projects:') !== -1;
    }

    private createRecord(line: string): ProjectRecord {
        if (!moment(line, 'YYYY-MM-DD').isValid()) {
            throw Error('invalid date')
        }
        return {
            day: new Date(line),
            projects: [],
        }
    }

    private parseMetadataProject(line: string): string {
        if (!line.match(METADATA_PROJECT_NAME_REGEX)) {
            throw Error('invalid metadata project');
        }
        const parsedProject = line.replace('-', '').trim();
        if (this.projects.indexOf(parsedProject) !== -1) {
            throw Error('duplicated metadata project');
        }
        return parsedProject;
    }

    private parseProject(line: string): EstimatedProject {
        const splitted = line.split(':');
        const [ name, mark ] = splitted;
        if (splitted.length !== 2 || this.currentRecord === undefined || !name.match(PROJECT_NAME_REGEX)) {
            throw new Error('invalid record')
        }
        if (this.currentRecord.projects.find((value) => value.name === name) !== undefined) {
            throw new Error('duplicated project record')
        }
        return {
            name: splitted[0],
            mark: this.parseMark(mark),
        };
    }

    private parseMark(value: string): Mark {
        // TODO: mark regex validate
        const trimmedValue = value.trim();
        if (trimmedValue === '-') {
            return undefined;
        }
        const mark = parseInt(value);
        if (Number.isNaN(mark) || mark < 1 || mark > 5) {
            throw new Error('invalid mark')
        }
        return mark as Mark;
    }
}
