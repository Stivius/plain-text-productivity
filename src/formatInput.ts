import chalk from "chalk";
import columnify from "columnify";
import dateFormat from "dateformat";
import prompt from "prompt";

function formatDate(date: Date) {
  return dateFormat(date, "yyyy-mm-dd");
}

export function formatInputForFile(date: Date, data: prompt.Properties) {
  return `${formatDate(date)}\n`
    .concat(
      Object.keys(data)
        .map((k) => {
          return `${k}:${data[k]}`;
        })
        .join("\n")
    )
    .concat("\n\n");
}

export function formatInputForConsole(
  date: Date,
  data: prompt.Properties
): string {
  const colorizeMark = (value: string) => {
    switch (value) {
      case "1":
        return chalk.rgb(150, 0, 0).bold(value);
      case "2":
        return chalk.rgb(255, 125, 125).bold(value);
      case "3":
        return chalk.rgb(255, 255, 75).bold(value);
      case "4":
        return chalk.rgb(125, 255, 125).bold(value);
      case "5":
        return chalk.rgb(0, 180, 0).bold(value);
    }
    return chalk.white(value);
  };

  return `${chalk.blue(formatDate(date))}\n`.concat(
    columnify(
      Object.keys(data).map((k) => ({
        project: chalk.yellow(k),
        mark: colorizeMark(data[k] as string),
      }))
    )
  );
}
