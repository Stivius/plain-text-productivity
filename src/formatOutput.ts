import chalk from 'chalk';
import columnify from 'columnify'

export function formatDataForFile(data) {
	return Object.keys(data).map((k) => {
		if (k !== 'Date') {
			return `${k}:${data[k]}`
		} else {
			return `${data[k]}`;
		}
	}).join('\n').concat('\n');
}

export function outputDataToConsole(data) {
	const colorizeMark = (value: string) => {
		switch (value) {
			case '1': return chalk.rgb(150, 0, 0).bold(value); 
			case '2': return chalk.rgb(255, 125, 125).bold(value); 
			case '3': return chalk.rgb(255, 255, 75).bold(value); 
			case '4': return chalk.rgb(125, 255, 125).bold(value); 
			case '5': return chalk.rgb(0, 180, 0,).bold(value); 
		}
		return chalk.white(value);
	};

	const DATE_KEY = 'Date';
	if (data.hasOwnProperty(DATE_KEY)) {
		console.log(`${chalk.blue(data[DATE_KEY])}`);
		delete data[DATE_KEY];
	}

	console.log(columnify(Object.keys(data).map((k) => ({
		project: chalk.yellow(k),
		mark: colorizeMark(data[k] as string),
	}))));
}

