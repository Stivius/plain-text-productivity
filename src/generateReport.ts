import chain, {filter, round, sumBy} from 'lodash'
import {FileData, ProductivityReport, ProductivityReportOptions} from './interfaces'

export function generateReport(data: FileData, options: ProductivityReportOptions): ProductivityReport {
    const groupedData = chain(data.records)
    .filter((r) => r.day >= options.from && r.day <= options.to)
    .map((r) => r.projects)
    .flatMap()
    .filter((p) => data.metadata.activeProjects.indexOf(p.name) !== -1)
    .groupBy((assesed) => assesed.name);
    const averageMarkByProject = groupedData.map((group, groupName) => {
        const groupSum = sumBy(group, (assesed) => assesed.mark);
        const filteredGroup = filter(group, (assesed) => assesed.mark !== undefined);
        return { name: groupName, mark: groupSum / filteredGroup.length };
    });
    const maxMarkFromAverages = averageMarkByProject.map(p => p.mark - 1).max();
    const items = averageMarkByProject.map((value) => {
        const mark = value.mark - 1;
        const MAX_MARK = 4;
        if (options.absolute) {
            return { name: value.name, productivity: round(mark / MAX_MARK, 4) };
        }
        return { name: value.name, productivity: round(mark / maxMarkFromAverages, 4) };
    }).value();
    return {
        items: items,
        options: options
    };
}
