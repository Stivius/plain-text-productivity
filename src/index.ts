import commandLineArgs from 'command-line-args';
import fs from 'fs';
import dateFormat from 'dateformat'
import {readRecordFromConsole} from './readRecordFromInput';
import chalk from 'chalk';
import {formatDataForFile, outputDataToConsole} from './formatOutput';
import {EstimatedProject, FileData, FILE_NAME, ProjectRecord} from './interfaces';
import {readRangeChoiceFromConsole} from './readRangeChoiceFromInput';
import {getRangeByType} from './getRangeByType';
import {FileParser} from './fileParser';
import {generateReport} from './generateReport';
import {outputReport} from './outputReport';

const commandsDefinitions = [
    { name: 'command', defaultOption: true },
]

const commands = commandLineArgs(commandsDefinitions, { stopAtFirstUnknown: true });

async function main() {
    try {
        const parser = new FileParser(FILE_NAME);
        const data: FileData = parser.parse();

        switch (commands.command) {
            // TODO: refactor output to console/file
            case 'add': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects);
                const date = dateFormat(new Date(), "yyyy-mm-dd");
                console.log(chalk.blue(date));
                fs.appendFileSync(FILE_NAME, date.concat('\n', formatDataForFile(enteredData), '\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'addp': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects, true);
                fs.appendFileSync(FILE_NAME, formatDataForFile(enteredData).concat('\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'lsp':
                const projects = data.records.map((r: ProjectRecord) => r.projects.map((p: EstimatedProject) => p.name));
            const uniqueProjects = Array.from(new Set(projects.flat()));
            const achivedProjects = uniqueProjects.filter((p) => !data.metadata.activeProjects.includes(p));
            console.log('All projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            console.log(chalk.gray(achivedProjects.join('\n')));
            break;

            case 'lspa':
                console.log('Active projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            break;

            case 'report': {
                const optionDefinitions = [
                    { name: 'absolute', type: Boolean, defaultValue: false },
                    { name: 'range-type', type: Number },
                    { name: 'from-date', type: String },
                    { name: 'to-date', type: String },
                ]
                const options = commandLineArgs(optionDefinitions, { argv: commands._unknown || [] });
                const getRange = (options: commandLineArgs.CommandLineOptions) => {
                    const rangeType = options['range-type'];
                    if (rangeType === undefined) {
                        return readRangeChoiceFromConsole();
                    } else {
                        // TODO: can be entered as string and converted to number
                        if (rangeType === 0) {
                            const from = new Date(options['from-date']);
                            const to = new Date(options['to-date']);
                            if (!from || !to) {
                                throw new Error('Date range is not specified or invalid');
                            }
                            return { from, to, choice: 0 };
                        }
                        return getRangeByType(rangeType);
                    }
                };

                const { from, to, choice } = await getRange(options);
                const productivityByProject = generateReport(data.records, from, to, options.absolute);
                outputReport(productivityByProject, choice, from, to, options.absolute);
                break;
            }
            default:
                console.error('Unknown command');
            break;
        }
    } catch (error) {
        console.error(error);
    }
}

main().catch((err) => console.log(err, 'Error while running'))
