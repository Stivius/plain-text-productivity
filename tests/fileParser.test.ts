import {FileParser} from '../src/fileParser'
import fs from 'fs'

jest.mock('fs')

const DATA = `
---
Projects:
- First
- Second
- Third
---
   
2022-05-22
First:1
Second:2
Third:3
Test:1 

2022-05-23
First:4
Second:5
Third:2
Test:- 
Test2:-

`;

// FIXME record should be parsed
const DATA_WITHOUT_METADATA = `
2022-05-22
First:1
Second:2
Third:3
Test:1`;

const DATA_ONLY_WITH_METADATA = `
---
Projects:
- First
- a1234
-abcd
---`;

const DATA_WITH_UNCLOSED_METADATA = `
---
Projects:
- First`;

// FIXME 2 records should be parsed
const DATA_WITHOUT_NEW_LINES = `
---
Projects:
- First
---
2022-05-22
First:1
2022-05-23
First:2`;

const DUPLICATED_METADATA = `
---
Projects:
- First
- First
---
2022-05-22
First:1

`;

const DUPLICATED_RECORD = `
---
Projects:
- First
---
2022-05-22
First:1
First:1

`;

const dataWithInvalidProject = (project: string) => `
---
Projects:
- First
${project}
--- `;

const dataWithInvalidDate = (date: string) => `
---
Projects:
- First
---
${date}
First:1

`;

const dataWithInvalidMark = (mark: string) => `
---
Projects:
- First
---
2022-05-22
First:${mark}

`;

const dataWithInvalidRecord = (record: string) => `
---
Projects:
- First
---
2022-05-22
${record}

`;

const FILE_NAME = 'sample.txt';

describe('fileParser', () => {
    const readFileSyncMocked = fs.readFileSync as jest.Mock;
    const parser = new FileParser(FILE_NAME);

    beforeEach(() => {
        jest.resetAllMocks();
    });

    test.each([
        DATA,
        DATA_WITHOUT_METADATA,
        DATA_ONLY_WITH_METADATA,
        DATA_WITH_UNCLOSED_METADATA,
        DATA_WITHOUT_NEW_LINES
    ])('parse valid data', (data: string) => {
        readFileSyncMocked.mockReturnValue(data);

        expect(parser.parse()).toMatchSnapshot();
    });

    test.each([
        DUPLICATED_METADATA,
        DUPLICATED_RECORD,
    ])('parse duplicated data', (data: string) => {
        readFileSyncMocked.mockReturnValue(data);

        expect(parser.parse.bind(parser)).toThrowError();
    });

    test('parse invalid file', () => {
        readFileSyncMocked.mockReturnValue('dummy');

        expect(parser.parse()).toMatchSnapshot();
    });

    // FIXME: throw exception
    test.each([
        '2022/05/22',
        '2022-99-22',
        '9999-05-22',
        '2022-05-99',
        '2022-02-30',
    ])('parse invalid date %p', (date: string) => {
        readFileSyncMocked.mockReturnValue(dataWithInvalidDate(date));

        expect(parser.parse.bind(parser)).toThrowError();
    });

    test.each([
        'dummy',
        '-1',
        '0',
        '6',
        Number.NEGATIVE_INFINITY,
        Number.POSITIVE_INFINITY,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_SAFE_INTEGER,
    ])('parse invalid mark %p', (mark: string) => {
        readFileSyncMocked.mockReturnValue(dataWithInvalidMark(mark));

        expect(parser.parse.bind(parser)).toThrowError();
    });

    // FIXME exception for invalid project
    test.each([
        'dummy',
        '12345',
        '- 1abcde',
        '- test/',
    ])('parse invalid project %p', (project: string) => {
        readFileSyncMocked.mockReturnValue(dataWithInvalidProject(project));

        expect(parser.parse.bind(parser)).toThrowError();
    });

    // FIXME exception for invalid project name
    test.each([
        'dummy',
        '1:1',
        '-t:1',
        't-:1',
        'test/:1',
    ])('parse invalid record %p', (record: string) => {
        readFileSyncMocked.mockReturnValue(dataWithInvalidRecord(record));

        expect(parser.parse.bind(parser)).toThrowError();
    });
});
