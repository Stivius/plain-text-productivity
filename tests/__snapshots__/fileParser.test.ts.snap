// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fileParser parse invalid file 1`] = `
Object {
  "metadata": Object {
    "projects": Array [],
  },
  "records": Array [],
}
`;

exports[`fileParser parse valid data 1`] = `
Object {
  "metadata": Object {
    "projects": Array [
      "First",
      "Second",
      "Third",
    ],
  },
  "records": Array [
    Object {
      "day": 2022-05-22T00:00:00.000Z,
      "projects": Array [
        Object {
          "mark": 1,
          "name": "First",
        },
        Object {
          "mark": 2,
          "name": "Second",
        },
        Object {
          "mark": 3,
          "name": "Third",
        },
        Object {
          "mark": 1,
          "name": "Test",
        },
      ],
    },
    Object {
      "day": 2022-05-23T00:00:00.000Z,
      "projects": Array [
        Object {
          "mark": 4,
          "name": "First",
        },
        Object {
          "mark": 5,
          "name": "Second",
        },
        Object {
          "mark": 2,
          "name": "Third",
        },
        Object {
          "mark": undefined,
          "name": "Test",
        },
        Object {
          "mark": undefined,
          "name": "Test2",
        },
      ],
    },
  ],
}
`;

exports[`fileParser parse valid data 2`] = `
Object {
  "metadata": Object {
    "projects": Array [],
  },
  "records": Array [
    Object {
      "day": 2022-05-22T00:00:00.000Z,
      "projects": Array [
        Object {
          "mark": 1,
          "name": "First",
        },
        Object {
          "mark": 2,
          "name": "Second",
        },
        Object {
          "mark": 3,
          "name": "Third",
        },
        Object {
          "mark": 1,
          "name": "Test",
        },
      ],
    },
  ],
}
`;

exports[`fileParser parse valid data 3`] = `
Object {
  "metadata": Object {
    "projects": Array [
      "First",
      "a1234",
      "abcd",
    ],
  },
  "records": Array [],
}
`;

exports[`fileParser parse valid data 4`] = `
Object {
  "metadata": Object {
    "projects": Array [
      "First",
    ],
  },
  "records": Array [],
}
`;

exports[`fileParser parse valid data 5`] = `
Object {
  "metadata": Object {
    "projects": Array [
      "First",
    ],
  },
  "records": Array [
    Object {
      "day": 2022-05-22T00:00:00.000Z,
      "projects": Array [
        Object {
          "mark": 1,
          "name": "First",
        },
      ],
    },
    Object {
      "day": 2022-05-23T00:00:00.000Z,
      "projects": Array [
        Object {
          "mark": 2,
          "name": "First",
        },
      ],
    },
  ],
}
`;
